src/sage/modules/free_module.py

```diff
@@ -6680,39 +6696,64 @@ def __init__(self, ambient, basis, check=True,
-            basis = self._echelonized_basis(ambient, basis)
+        MAT = sage.matrix.matrix_space.MatrixSpace(
+            R_coord, len(basis), ambient.degree(), sparse=ambient.is_sparse())
+        
+        A = MAT(basis)
+
+        # Rank computation is expected to be fast unlike echelon form computation
+        # TODO: Avoid rank computation during module construction
+        rank = A.rank()
 
         # Adapted from Module_free_ambient.__init__
         from sage.categories.modules_with_basis import ModulesWithBasis
         modules_category = ModulesWithBasis(R.category()).FiniteDimensional()
         try:
-            if R.is_finite() or len(basis) == 0:
+            if R.is_finite() or rank == 0:
                 modules_category = modules_category.Enumerated().Finite()
         except (ValueError, TypeError, AttributeError, NotImplementedError):
             pass
         modules_category = modules_category.Subobjects()
         category = modules_category.or_subcategory(category, join=True)
 
         FreeModule_generic_pid.__init__(self, base_ring=R, coordinate_ring=R_coord,
-                                        rank=len(basis), degree=ambient.degree(),
+                                        rank=rank, degree=ambient.degree(),
                                         sparse=ambient.is_sparse(), category=category)
+
+        if echelonize and not already_echelonized:
+            matrix = self._echelonized_basis(ambient, basis)
+            assert rank == matrix.nrows()
+            basis = matrix.rows()
+
         C = self.element_class
         w = [C(self, x.list(), coerce=False, copy=False) for x in basis]
         self.__basis = basis_seq(self, w)
 
-        if echelonize or already_echelonized:
-            self.__echelonized_basis = self.__basis
-        else:
-            if echelonized_basis is None:
-                echelonized_basis = self._echelonized_basis(ambient, basis)
-            w = [C(self, x.list(), coerce=False, copy=True)
-                 for x in echelonized_basis]
-            self.__echelonized_basis = basis_seq(self, w)
-        if check and len(basis) != len(self.__echelonized_basis):
-            raise ValueError("The given basis vectors must be linearly "
-                             "independent.")
+        has_echelonized_basis = echelonize or already_echelonized or echelonized_basis
+
+        if has_echelonized_basis:
+            if echelonize or already_echelonized:
+                self.__echelonized_basis = self.__basis
+            elif echelonized_basis:
+                w = [C(self, x.list(), coerce=False, copy=False) for x in echelonized_basis]
+                self.__echelonized_basis = basis_seq(self, w)
+
+                # TODO: Validate span of echelonized basis with provided basis
+                if check and len(self.__echelonized_basis) != rank:
+                    raise ValueError("The given echelonized basis vectors are not correct.")
```

Comment:
```suggestion
                    raise ValueError("the given echelonized basis vectors are not correct")
```
This is to follow Python conventions. That said, why is it not correct? This is not a useful error message. Say that the number of basis vectors does not match the rank.